### PR TITLE
Merge branch-command-history to branch-delivery

### PR DESCRIPTION
**Pull request by Choo-Xing-Yu**
Implemented History

CommandBox now has Window's command prompt Command History feature in the sense that:
1) Pressing arrow up key cycles through your past commands (if any)
2) Pressing arrow down key cycles through your finished commands (if any)

This implementation aligns with aiding our user's fast typist perk

### CODE REVIEW

File: src/main/java/seedu/address/history/CommandHistory.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    private boolean isLimitReached() {
+        return commandHistory.size() == lengthLimit;
+    }
+
+    private boolean isWithinLengthLimit() {
+        return currentCommandIndex >= 0 && currentCommandIndex < lengthLimit;
+    }
+
+    private boolean isAbleToReturnNextCommand() {
+        return currentCommandIndex < commandHistory.size() - 1 && isWithinLengthLimit();
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/history/CommandHistory.java]: Could be more SLAP like
`return hasNextCommand() && isWithinLengthLimit();`

File: src/main/java/seedu/address/history/CommandHistory.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    private boolean isAbleToReturnPreviousCommand() {
+        return currentCommandIndex > 0 && isWithinLengthLimit();
+    }
+
+    private boolean isCurrentCommandIndexZero() {
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/history/CommandHistory.java:79]: Is this abstraction really necessary?

File: src/main/java/seedu/address/history/CommandHistory.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            this.currentCommandIndex++;
+            this.commandHistory.add(command);
+        }
+    }
+
+    @Override
+    public Optional<String> previousCommand() {
+        if (!hasReturnedCurrentCommandBefore) {
+            hasReturnedCurrentCommandBefore = true;
+            return currentCommand();
```
Comment (saad-haider-pp) [inline src/main/java/seedu/address/history/CommandHistory.java:112]: I'm a little confused as to what the role of hasReturnedCurrentCommandBefore is. Could you add a comment to the code explaining a little what is going on

Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/history/CommandHistory.java:112]: Take a look now! added comments

File: src/main/java/seedu/address/history/CommandHistory.java (hunk 1 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/history/CommandHistory.java]: Fixed! take a look!

File: src/main/java/seedu/address/history/CommandHistory.java (hunk 2 above)
Comment (Choo-Xing-Yu) [reply src/main/java/seedu/address/history/CommandHistory.java:79]: to adhere to similar level of abstraction in the code

Comment (saad-haider-pp) [reply src/main/java/seedu/address/history/CommandHistory.java:79]: Fair enough

File: src/main/java/seedu/address/history/CommandHistory.java (hunk 3 above)
Comment (saad-haider-pp) [reply src/main/java/seedu/address/history/CommandHistory.java:112]: Looks good. LGTM!. Really good job on figuring out the logic

**Pull request merged**